### PR TITLE
[16.0][IMP] mgmtsystem_nonconformity_quality_control_oca: auto-create nonco…

### PR DESCRIPTION
**Pull request by quirino95**
Added new flag "Create Nonconformity" on products.

When an inspection related to a picking fails, related nonconformities are automatically created for all products having the flag active.

### CODE REVIEW

**Review by HekkiMelody — APPROVED** [review]
Code review, LGTM

**Review by marcos-mendez — DISMISSED** [review]
**Automated Review -- Tests Failed**

### 1. **Root Cause**

The test failure is caused by a **database connection error** during the Odoo startup, which is unrelated to the code changes in this PR. This is likely due to an infrastructure or environment issue in the CI/runner setup (e.g., PostgreSQL not running or misconfigured).

---

### 2. **Suggested Fix**

There is **no code fix needed** for the test failure itself, as it's a runtime environment issue. However, to ensure robustness of the module:

- In `qc_inspection.py`, line 53: The condition `if self.state == "failed" and self.product_id.create_nonconformity:` should be safe, but to avoid potential errors if `self.product_id` is `False`, add a null check:
  ```python
  if self.state == "failed" and self.product_id and self.product_id.create_nonconformity:
  ```

---

### 3. **Additional Code Issues**

- **Missing null check in `action_approve`**: If `self.product_id` is `False` (e.g., inspection is not linked to a product), the code will raise an AttributeError. This should be guarded.
  - File: `mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py`
  - Line: 53

---

### 4. **Test Improvements**

The current tests cover basic functionality, but could be improved with:

- **TransactionCase** tests using `SavepointCase` to test edge cases:
  - Test with a product that has `create_nonconformity = True` but inspection is not linked to a product.
  - Test with inspection state not set to "failed" — ensure no NC is created.
  - Test that the created NC has correct values for `qc_inspection_id`, `name`, `description`, etc.

Example test case to add:
```python
def test_create_nonconformity_with_no_product(self):
    self.inspection2.write({"state": "failed", "object_id": False})
    self.inspection2.action_approve()
    nc = self.env["mgmtsystem.nonconformity"].search([("qc_inspection_id", "=", self.inspection2.id)])
    self.assertFalse(nc)
```

Use **tagged tests** (e.g., `@tag('post_install')`) for tests that require data setup or module dependencies.

---

### Summary

- The test failure is **environment-related**, not code-related.
- Add a null check for `self.product_id` in `action_approve`.
- Improve test coverage using `SavepointCase` and `TransactionCase` with edge cases.

---
### ⏰ PR Aging Alert

This PR by @quirino95 has been open for **103 days** (3 months).
💤 Last activity was 41 days ago.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! I found some test failures on this PR and left detailed feedback above. I am happy to discuss or help debug. In the meantime, if any of you get a chance, I would appreciate a look at my open PR(s):

**My open PRs across OCA:**
- [server-tools#3554 [MIG] datetime_formatter: Migration to 18.0](https://github.com/OCA/server-tools/pull/3554)
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward. Thank you!

---
_Environment via [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer): Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

File: mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py [hunk 1]
```diff
@@ -48,3 +48,24 @@ def action_view_nonconformities(self):
                 "default_company_id": self.company_id.id,
             }
         return action
+
+    def action_approve(self):
+        res = super().action_approve()
+
+        if self.state == "failed" and self.product_id.create_nonconformity:
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py:55]: **issue**: if `self` is a recordset, accessing `self.state` raises the error
```
  [...]
  File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 1154, in __get__
    record.ensure_one()
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 5222, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
```
please loop on `self` and call `create_nonconformity` only on appropriate inspections.
A test checking approval of multiple inspections would also prevent a regression.

File: mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -5,3 +5,4 @@ This module contains some new features for Management System modules.
 Nonconformity (NC)
 
 * Quality Control Inspection: add a field to link a specific quality control inspection.
+* Activating "Create Nonconformity" flag on products, for every failed inspection, a new Nonconformity will be opened.
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst:8]: **suggestion**: Since this is a configuration, it could be also mentioned in a `CONFIGURE` fragment.

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:1]: **praise**: Thanks for covering this improvement with tests 🙏

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py [hunk 3]
```diff
@@ -56,3 +65,31 @@ def test_action_view_nonconformities(self):
             nc_ids.append(nc.id)
 
         self.assertEqual(nc_ids, action["domain"][0][2])
+
+    def test_no_create_nonconformity(self):
+        """
+        Test Nonconformity creation when inspection fails
+        if relative flag is off
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:72]: **note**: I don't understand this docstring: this says it is testing nonconformity creation, but there is no nonconformity record being created in the test; on the contrary: the test is actually checking that no nonconformity record is being created.
Could you please clarify the docstring?

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py [hunk 4]
```diff
@@ -56,3 +65,31 @@ def test_action_view_nonconformities(self):
...
+        nc = self.env["mgmtsystem.nonconformity"].search(
+            [("qc_inspection_id", "=", self.inspection2.id)]
+        )
+        self.assertFalse(nc)
+
+    def test_create_nonconformity(self):
+        """
+        Test Nonconformity doesn't create when inspection fails
+        if relative flag is on
+        """
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:86]: **suggestion**: This test actually checks that the the nonconfromity record is created.

```suggestion
        """
        Test Nonconformity is created when inspection fails
        if relative flag is on
        """
```

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py [hunk 5]
```diff
@@ -56,3 +65,31 @@ def test_action_view_nonconformities(self):
...
+        self.product.create_nonconformity = True
+        self.inspection2.write({"state": "failed"})
+        self.inspection2.action_approve()
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:89]: **suggestion**: Please check there are no nonconformities before approval, otherwise this test might pass for a false positive.

File: mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record model="ir.ui.view" id="product_template_form_view">
+        <field name="name">product.template.common.qc</field>
```
Comment (SirPyTech) [inline mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml:5]: **note**: The name of this view should briefly describe what it's doing, this is just copy/pasted from the inherited view.
Please mention something about nonconformities

**Review by SirPyTech — CHANGES_REQUESTED** [review]
Thanks for the PR!
I only reviewed the code and have a few suggestions, also not the **issue** that could raise an exception

File: mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml (hunk 6 above)
Comment (anusriNPS) [reply mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml]: Updated name which includes module name

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py (hunk 5 above)
Comment (anusriNPS) [reply mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:94]: Included no nonconformities check before marking inspection as failed as per suggestion

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py (hunk 3 above)
Comment (anusriNPS) [reply mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py]: Docstring is corrected as per testcase implementation

File: mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst (hunk 2 above)
Comment (anusriNPS) [reply mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst]: Updated documentation as per suggestion

File: mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py (hunk 1 above)
Comment (anusriNPS) [reply mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py]: Updated loop on self and introduced testcase to check multiple inspections on action_approve.

File: mgmtsystem_nonconformity_maintenance/views/mgmtsystem_nonconformity.xml
Comment (SirPyTech) [inline mgmtsystem_nonconformity_maintenance/views/mgmtsystem_nonconformity.xml:1]: **chore**: These changes to a different module actually do nothing, I think you didn't want to add them, please roll them back

File: mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py (hunk 1 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/models/qc_inspection.py]: :+1: Thanks!

File: mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst (hunk 2 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/readme/DESCRIPTION.rst]: :+1:

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py (hunk 3 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py]: :+1:

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py (hunk 4 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:87]: :+1:

File: mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py (hunk 5 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/tests/test_inspection.py:94]: :+1:

File: mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml (hunk 6 above)
Comment (SirPyTech) [reply mgmtsystem_nonconformity_quality_control_oca/views/product_template.xml]: :+1:

**Review by SirPyTech — CHANGES_REQUESTED** [review]
I reviewed the code and tried it locally, looks good 👍
Please just remove the extra changes to another module

**Comment by SirPyTech** [comment]
@marcos-mendez please dismiss the generated review so we can see when actual reviews get this approved

File: mgmtsystem_nonconformity_maintenance/views/mgmtsystem_nonconformity.xml
Comment (anusriNPS) [reply mgmtsystem_nonconformity_maintenance/views/mgmtsystem_nonconformity.xml:1]: Rolled them back. Thank you for catching it.

Comment (SirPyTech) [reply mgmtsystem_nonconformity_maintenance/views/mgmtsystem_nonconformity.xml:1]: :+1:

**Review by SirPyTech — APPROVED** [review]

**Comment by SirPyTech** [comment]
@OCA/management-systems-maintainers please dismiss the generated review from @marcos-mendez so we can see this is now approved, thanks!

**Review by max3903 — APPROVED** [review]
## Review

Looks good on current HEAD. Prior feedback from @SirPyTech was addressed (recordset loop + `product_id` guard, docs, tests, unrelated module rollback). CI is green.

### What works
- Product flag `create_nonconformity` + create NC on `action_approve` when inspection failed
- Safe for multi-record `action_approve` and missing product
- Tests cover flag off / flag on / no product / mixed recordset
- Readme CONFIGURE / USAGE / DESCRIPTION updated

### Non-blocking nits (optional follow-up)
1. Prefer explicit M2M command for clarity:
   `"origin_ids": [(4, self.env.ref("mgmtsystem_nonconformity.nc_origin_qc").id)]`
2. Translate the hard-coded description with `_()`.
3. Add `self.ensure_one()` in `create_nonconformity()`.
4. `product_template.py` is missing a copyright header.
5. PR body says NCs for “all products” on a picking; implementation creates one NC for `inspection.product_id` only — wording mismatch only.
6. `USAGE.rst`: add a blank line after the “NC Automatic Creation for Products” heading so the bullet renders correctly in RST.

### Housekeeping
- Dismissed stale @marcos-mendez automated “changes requested” so the PR can show as approved.
- Removed outdated `needs fixing` (CI is green).

**Approve** — ready to merge once OCA age/review policy is satisfied (already ≥5 days; now ≥2 human approvals on latest intent).

**Comment by marcos-mendez** [comment]
sorry @SirPyTech just saw it now

**Comment by anusriNPS** [comment]
> ## Review
> 
> Looks good on current HEAD. Prior feedback from @SirPyTech was addressed (recordset loop + `product_id` guard, docs, tests, unrelated module rollback). CI is green.
> ### What works
> 
>     * Product flag `create_nonconformity` + create NC on `action_approve` when inspection failed
> 
>     * Safe for multi-record `action_approve` and missing product
> 
>     * Tests cover flag off / flag on / no product / mixed recordset
> 
>     * Readme CONFIGURE / USAGE / DESCRIPTION updated
> 
> 
> ### Non-blocking nits (optional follow-up)
> 
>     1. Prefer explicit M2M command for clarity:
>        `"origin_ids": [(4, self.env.ref("mgmtsystem_nonconformity.nc_origin_qc").id)]`
> 
>     2. Translate the hard-coded description with `_()`.
> 
>     3. Add `self.ensure_one()` in `create_nonconformity()`.
> 
>     4. `product_template.py` is missing a copyright header.
> 
>     5. PR body says NCs for “all products” on a picking; implementation creates one NC for `inspection.product_id` only — wording mismatch only.
> 
>     6. `USAGE.rst`: add a blank line after the “NC Automatic Creation for Products” heading so the bullet renders correctly in RST.
> 
> 
> ### Housekeeping
> 
>     * Dismissed stale @marcos-mendez automated “changes requested” so the PR can show as approved.
> 
>     * Removed outdated `needs fixing` (CI is green).
> 
> 
> **Approve** — ready to merge once OCA age/review policy is satisfied (already ≥5 days; now ≥2 human approvals on latest intent).

 Updated non blocking points (1,2,3,6). rst file alignment looks better now

**Review by max3903 — APPROVED** [review]
## Re-review (new HEAD `027d421`)

Thanks @anusriNPS — the optional nits from the previous review are addressed:

1. ✅ `origin_ids` uses `[(4, xmlid.id)]`
2. ✅ description wrapped with `_()`
3. ✅ `ensure_one()` in `create_nonconformity()`
4. ⬜ `product_template.py` still has no copyright header (optional)
5. ⬜ PR description still mentions “all products” vs single `product_id` (docs only)
6. ✅ `USAGE.rst` blank line / bullet formatting fixed

CI is green again. **Re-approve** on the new commit — still good to merge (`/ocabot merge patch`).
